### DISCUSSION TITLE
Unable to determine a compatible waveformData format

### DISCUSSION THREAD

**Discussion by Grumpier** [bbc/peaks.js#466] [Q&A] [answered]
I'm trying to serve a precomputed json waveform to Peaks.init using the waveformData option and get the above error message. i have no problem if I use dataUri to point to the same file on the server. Here's my code: 

```
      const track_id = document.getElementById("track_id").dataset.id;
      async function getJson() {
        const wave = await fetch(${origin}/wave/${track_id});
        return wave.json();
      }
    
      (async function(Peaks) {
        const wave = await getJson();
        console.log(wave);
        var options = {
          containers: {
            zoomview: document.getElementById('zoomview-container'),
            overview: document.getElementById('overview-container') },
          zoomview: {waveformColor: 'rgba(0, 0, 255, 1)'},
          mediaElement: document.getElementById('audio'),
          waveformData: { json: wave },
          keyboard: true,
          pointMarkerColor: '#006eb0',
          showPlayheadTime: true,
          emitCueEvents: true
        };
```

Here's what the console.log of the waveform looks like:
![Screen Shot 2022-10-09 at 11 20 31](https://user-images.githubusercontent.com/12978695/194752530-2b90d18f-50b4-4ea6-b82d-778951bdcd1f.png)

Any ideas of what I'm doing wrong? Thanks!

**Comment by chrisn** [comment]
I don't see anything immediately wrong. The code that produces the error is [here](https://github.com/bbc/peaks.js/blob/v2.0.6/src/waveform-builder.js#L235), does that suggest anything?

**Reply by Grumpier** [reply]
Thanks chrisn. This helped. The `isObject(waveformData.json)` test is failing because the wave value I am passing to options is apparently still wrapped in a promise. It's odd, because `console.log(options.waveformData.json)` properly displays it as a json object. I've tried wrapping all of the peaks function inside of `getJson().then`, since I thought the promise should then be unwrapped, but it still fails. I'm not a strong enough JS programmer to know exactly how to resolve this. Do you have any suggestion?

**Comment by chrisn** [comment, marked as answer]
I can only think that there's some difference between the code you posted here and what is in your project.

Because you're using `const wave = await getJson();` then `wave` will contain the data after the promise resolves.

If `options.waveformData.json` were a promise, you would see a different error: `WaveformData.create(): Unknown data format`

**Reply by Grumpier** [reply]
You're right! The promise is being resolved. The problem was at the server end. I was using 

`fs.readFile( fileName, "utf-8", function(err, data) {
   res.json(data) })` 

which fails. I switched to a simple

`res.sendFile (fileName)`

and it now works. Thanks for your help!

**Reply by chrisn** [reply]
Thanks! It sounds like the data was being stringified as JSON too many times.
